pkg/controller/orchestrator/orchestrator_reconcile.go

```diff
@@ -0,0 +1,478 @@
+		}
+
+		// check if cluster is read only
+		isReadOnly = isReadOnly && node.ReadOnly
+	}
+
+	// set master node Master condition on True
+	master := insts.DetermineMaster()
+	if master == nil {
+		log.Error(fmt.Errorf("error detemining master"), "error acquiring master name", "instances", insts)
+	} else {
+		ou.updateNodeCondition(master.Key.Hostname, api.NodeConditionMaster, core.ConditionTrue)
+	}
+
+	// set cluster ReadOnly condition
+	if isReadOnly {
+		ou.cluster.UpdateStatusCondition(api.ClusterConditionReadOnly,
+			core.ConditionTrue, "ClusterReadOnlyTrue", "cluster is in read only")
+	} else {
+		ou.cluster.UpdateStatusCondition(api.ClusterConditionReadOnly,
+			core.ConditionFalse, "ClusterReadOnlyFalse", "cluster is writable")
+	}
+}
+
+// registerUnregisterNodesInOrc is the functions that tries to register
+// unregistered nodes and to remove nodes that does not exists.
+func (ou *orcUpdater) registerUnregisterNodesInOrc(instances InstancesSet) InstancesSet {
+	var (
+		// hosts that should be discovered
+		shouldBeDiscovered []string
+		// list of instances from orchestrator that should be removed
+		shouldBeRemoved []string
+		// the list of instances that are in orchestrator and k8s
+		instancesFiltered InstancesSet
+	)
+
+	for i := 0; i < ou.cluster.Status.ReadyNodes; i++ {
+		host := ou.cluster.GetPodHostname(i)
+		if inst := instances.GetInstance(host); inst == nil {
+			// host is not present into orchestrator
+			// register new host into orchestrator
+			shouldBeDiscovered = append(shouldBeDiscovered, host)
+		} else {
+			// this instance is present in both k8s and orchestrator
+			instancesFiltered = append(instancesFiltered, *inst)
+		}
+	}
+
+	// remove all instances from orchestrator that does not exists in k8s
+	for _, inst := range instances {
+		if i := instancesFiltered.GetInstance(inst.Key.Hostname); i == nil {
+			shouldBeRemoved = append(shouldBeRemoved, inst.Key.Hostname)
+		}
+	}
+
+	ou.discoverNodesInOrc(shouldBeDiscovered)
+	ou.forgetNodesFromOrc(shouldBeRemoved)
+
+	return instancesFiltered
+}
+
+func (ou *orcUpdater) discoverNodesInOrc(hosts []string) {
+	for _, host := range hosts {
+		if err := ou.orcClient.Discover(host, mysqlPort); err != nil {
+			log.Error(err, "failed to register host with orchestrator", "host", host)
+		}
+	}
+}
+
+func (ou *orcUpdater) forgetNodesFromOrc(hosts []string) {
+	for _, host := range hosts {
+		if err := ou.orcClient.Forget(host, mysqlPort); err != nil {
+			log.Error(err, "failed to forget host with orchestrator", "host", host)
+		}
+	}
+}
+
+func (ou *orcUpdater) getRecoveriesToAck(recoveries []orc.TopologyRecovery) []orc.TopologyRecovery {
+	toAck := []orc.TopologyRecovery{}
+
+	if len(recoveries) == 0 {
+		return toAck
+	}
+
+	i, found := condIndexCluster(ou.cluster, api.ClusterConditionReady)
+	if !found || ou.cluster.Status.Conditions[i].Status != core.ConditionTrue {
+		log.Info("cluster not ready for acknowledge", "cluster", ou.cluster)
```

Comment:
`cluster not ready for acknowledge` -> `skip acknowledging recovery since cluster is not ready yet`